internal/toolexec/processors/package.go

```diff
@@ -0,0 +1,227 @@
+// This is aimed at library packages that don't yield and importcfg.link in their b001
+// compilation subtree
+func PreparePackage(importPath, pkgDir string, buildFlags ...string) (*PkgRegister, error) {
+	// 1 - Build pkg
+	log.Printf("====> Building %s", importPath)
+	wDir, err := utils.GoBuild(pkgDir, buildFlags...)
+	if err != nil {
+		return nil, err
+	}
+
+	pkgReg := newPkgReg(importPath, wDir)
+
+	// 2 - Fetch and combine all dependencies
+	log.Printf("====> Building pkg register for %s", importPath)
+	filepath.WalkDir(wDir, func(path string, d fs.DirEntry, err error) error {
+		if d.IsDir() || d.Name() != "importcfg" {
+			return nil
+		}
+		file, err := os.Open(path)
+		if err != nil {
+			return err
+		}
+		pkgReg.Combine(pkgRegiterFromImportCfg(file))
+		defer file.Close()
+		return nil
+	})
+
+	return &pkgReg, err
+}
+
+type PackageInjector struct {
+	importPath string
+	pkgDir     string
+	buildFlags []string
+}
+
+// NewPackageInjector creates a new package injector
+func NewPackageInjector(importPath, pkgDir string, flags ...string) PackageInjector {
+	return PackageInjector{
+		importPath: importPath,
+		pkgDir:     pkgDir,
+		buildFlags: flags,
+	}
+}
+
+// ProcessCompile visits a compile command, compiles the injected package
+// and includes the package dependency in the target package's importcfg
+func (i *PackageInjector) ProcessCompile(cmd *proxy.CompileCommand) {
+	if cmd.Stage() != "b001" {
+		return
+	}
+	log.Printf("[%s] Injecting %s at compile", cmd.Stage(), i.importPath)
+	// 1 - Build the package
+	pkgReg, err := PreparePackage(i.importPath, i.pkgDir, i.buildFlags...)
+	utils.ExitIfError(err)
+	state := State{
+		Deps: map[string]PkgRegister{i.importPath: *pkgReg},
+	}
+
+	// 2 - Add pkg dependency in importcfg
+	log.Printf("====> Injecting %s in final importcfg", i.importPath)
+	err = filepath.WalkDir(filepath.Dir(cmd.Flags.Output), func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			log.Printf("err at entry: %v", err)
+			return err
+		}
+		if d.IsDir() || d.Name() != "importcfg" {
+			return nil
+		}
+
+		file, err := os.OpenFile(path, os.O_APPEND|os.O_RDWR|os.O_CREATE, 0640)
+		if err != nil {
+			log.Printf("err browsing file dir: %v", err)
+			return err
+		}
+		defer file.Close()
+		str := fmt.Sprintf("packagefile %s=%s/b001/_pkg_.a", i.importPath, pkgReg.BuildDir)
+		_, err = file.WriteString(str)
+		return err
+	})
+
+	// 3 - Save state to disk for the link invocation (separate process)
+	utils.ExitIfError(state.SaveToFile(ddStateFilePath))
+	log.Printf("====> Saved state to %s", ddStateFilePath)
+}
+
+func (i *PackageInjector) ProcessLink(cmd *proxy.LinkCommand) {
+	if cmd.Stage() != "b001" {
+		return
+	}
+	log.Printf("[%s] Injecting %s at link", cmd.Stage(), i.importPath)
+
+	// 1 - Read state from disk (created by ProcessCompile step)
+	log.Printf("====> Reading state from %s", ddStateFilePath)
+	state, err := StateFromFile(ddStateFilePath)
+	defer os.Remove(ddStateFilePath)
```

Comment:
Maybe we should make it configurable? A mode where we remove ASM byproducts asap seems like something that we may want to allow. Maybe make it the default mode add an opt-in option to save the state?

Reply:
If these are _large_ objects, then yes, it might be a good thing to allow cleaning them up earlier rather than later...

My perspective on this is that the work tree will be cleaned up by the go toolchain unless `-work` is passed, and that would be the mechanism to allow the user to choose to keep intermediate files & directories

Reply:
Fair enough. Removing the line 👍